### ISSUE TITLE
Output new snapshot ID whenever a command causes an existing ID to change

### ISSUE THREAD

**Issue by maltem-za** [restic/restic#5137]
Output of `restic version`
--------------------------

restic 0.17.1 compiled with go1.23.1 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

For example when adding tags to a snapshot, it would be helpful if the output returned the resulting new snapshot ID.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

Finding the same snapshot again after its ID has changed. Having to list snapshots again to find the new ID, and potentially having to compare using other metadata fields to try to ensure that it's the same one.

Sometimes one might make a note of a specific ID because the snapshot was created for a specific purpose (eg: before an upgrade), or because it was determined to be the one needed out of a set of older ones. (Yes, one could also use tags to mark these, but that's not really the point.)

Perhaps it would also be possible to (optionally) list all previous IDs, although I think only the most recent previous one is recorded.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

It makes me happy every time it Just Works™, which is almost always.

**Comment by MichaelEischer** [comment]
> For example when adding tags to a snapshot, it would be helpful if the output returned the resulting new snapshot ID.

Good idea. If someone wants to make the necessary changes, that feature should be rather easy to add (just look at the existing debug logs in the `tag` command).

> Perhaps it would also be possible to (optionally) list all previous IDs, although I think only the most recent previous one is recorded.

A snapshot contains the original ID in its `original` field. Other IDs aren't stored, but are probably also not particularly useful.

**Comment by konidev20** [comment]
@MichaelEischer I can take a look at this. 

Since, multiple snapshots can change. I am thinking if the text output be something like:
```
$ restic -r ~/test-repo tag --set first
create exclusive lock for repository
enter password for repository: 
repository 27810913 opened (version 2, compression level auto)
modified tags on 2 snapshots
Old Snapshot: 7209dfaf, New Snapshot: da853b15
Old Snapshot: abcbf9e1, New Snapshot: 070dfb12
```

and currently, there is no JSON output for this command, so it can be something like:
```
{
  "modified_snapshots": [
    {
        "old_snapshotID": "7209dfaf",
        "new_snapshotID": "da853b15"
    },
    {
        "old_snapshotID": "abcbf9e1",
        "new_snapshotID": "070dfb12"
    }
  ]
}
```

**Comment by maltem-za** [comment]
Great that this was implemented for the `tag` command, thanks guys!

As this is now closed I've updated the title to be more accurate in relation to the merged PR. It was initially "a command" as there are other commands where this could be useful, e.g.: `copy` and `rewrite` - perhaps it would be worth creating separate issues for those?

**Comment by MichaelEischer** [comment]
> `copy` and `rewrite` - perhaps it would be worth creating separate issues for those?

Sure, feel free to open corresponding issues. For the `copy` command there is already an older issue to add JSON output, it might make sense to add a note there https://github.com/restic/restic/issues/3324

**Comment by maltem-za** [comment]
@konidev20 Maybe I'm missing something, but I just added a tag to a snapshot with 0.18 and it only printed the no. of snapshots modified?

**Comment by konidev20** [comment]
> @konidev20 Maybe I'm missing something, but I just added a tag to a snapshot with 0.18 and it only printed the no. of snapshots modified?

I think I found the issue too.

Let me have a look this weekend and post the RCA here.

**Comment by konidev20** [comment]
@maltem-za , itseems to be working for me. Could you describe in more detail what you're trying to do? the command and the output. Also, if there were already some tags associated with it.

**Comment by konidev20** [comment]
```
restic -r ~/test-repo tag --add lmo,qpr,112 65978a8c --json
create exclusive lock for repository
enter password for repository: 
{"message_type":"changed","old_snapshot_id":"65978a8cca149cc55eeb5e46814f5ad52e82714e1c611f62d35fe09790c4da30","new_snapshot_id":"249d8423c729ea12a6dd5b716081e9c5d86edf2b097ce35bb346e282b954b77b"}
{"message_type":"summary","changed_snapshots":1}
```

**Comment by konidev20** [comment]
Oh got it, after a bit more of fiddling. 

Check out this review comment: https://github.com/restic/restic/pull/5144#discussion_r1864872384

in JSON mode, it prints the original snapshot and new snapshot ID.

In text mode, it prints it only in `--verbose` mode.
